src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.cs

```diff
@@ -773,5 +819,28 @@ public void MarkSyntaxAnalysisComplete(SyntaxTree tree, IEnumerable<DiagnosticAn
                 GetAnalyzerState(analyzer).MarkSyntaxAnalysisComplete(tree);
             }
         }
+
+        /// <summary>
+        /// Marks the given tree as fully syntactically analyzed for the unprocessed analyzers in the given analysisScope.
+        /// </summary>
+        public void MarkSyntaxAnalysisCompleteForUnprocessedAnalyzers(
+            SyntaxTree tree,
+            AnalysisScope analysisScope,
+            PooledHashSet<DiagnosticAnalyzer> processedAnalyzers)
+        {
+            Debug.Assert(processedAnalyzers.All(analysisScope.Contains));
+            if (analysisScope.Analyzers.Length == processedAnalyzers.Count)
+            {
+                return;
+            }
+
+            foreach (var analyzer in analysisScope.Analyzers)
+            {
+                if (!processedAnalyzers.Contains(analyzer))
+                {
+                    MarkSyntaxAnalysisComplete(tree, analyzer);
+                }
+            }
+        }
```

Comment:
>} [](start = 8, length = 1)

Consider extracting a helper method for use in all three of these methods: `MarkAnalysisCompleteForUnprocessedAnalyzers<T>(AnalysisScope scope, HashSet<DiagnosticAnalyzer> processedAnalyzers, Action<AnalysisState, T> markComplete, T arg)`

Reply:
Thanks, done!